src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,201 @@
+                    unlink(x);
+                    return true;
+                }
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean isElementIndex(int index) {
+        return index >= 0 && index < size;
+    }
+
+    private boolean isPositionIndex(int index) {
+        return index >= 0 && index <= size;
+    }
+
+    private String outOfBoundsMsg(int index) {
+        return "Index: " + index + ", Size: " + size;
+    }
+
+    private void checkPositionIndex(int index) {
+        if (!isPositionIndex(index)) {
+            throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    private void checkElementIndex(int index) {
+        if (!isElementIndex(index)) {
+            throw new IndexOutOfBoundsException(outOfBoundsMsg(index));
+        }
+    }
+
+    void linkBefore(T e, Node<T> succ) {
+        final Node<T> pred = succ.prev;
+        final Node<T> newNode = new Node<>(pred, e, succ);
+        succ.prev = newNode;
+        if (pred == null) {
+            first = newNode;
+        } else {
+            pred.next = newNode;
+        }
+        size++;
+    }
+
+    Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            Node<T> x = first;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            Node<T> x = last;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
+    }
+
+    void linkLast(T e) {
+        final Node<T> l = last;
+        final Node<T> newNode = new Node<>(l, e, null);
+        last = newNode;
+        if (l == null) {
+            first = newNode;
+        } else {
+            l.next = newNode;
+        }
+        size++;
+    }
+
+    T unlink(Node<T> x) {
+        final T element = x.value;
+        final Node<T> next = x.next;
+        final Node<T> prev = x.prev;
+
+        if (prev == null) {
+            first = next;
+        } else {
+            prev.next = next;
+            x.prev = null;
+        }
+
+        if (next == null) {
+            last = prev;
+        } else {
+            next.prev = prev;
+            x.next = null;
+        }
+
+        x.value = null;
+        size--;
+        return element;
+    }
+
+    private void linkAllLast(List<T> list) {
+        Node<T> lastNode = last;
+        for (int i = 0; i < list.size(); i++) {
+            Node<T> newNode = new Node<>(lastNode, list.get(i), null);
+            if (lastNode == null) {
+                first = newNode;
+            } else {
+                lastNode.next = newNode;
+            }
+            lastNode = newNode;
+        }
+        last = lastNode;
+        size += list.size();
+    }
```

Comment:
The method name `linkAllLast` is not very descriptive. Consider renaming it to something more meaningful, like `addAllAtEnd` or `linkAllNodesAtEnd`.